dashboard/src/components/FunctionList/FunctionDeployButton.tsx

```diff
@@ -0,0 +1,241 @@
+import * as Moniker from "moniker-native";
+import * as React from "react";
+import AceEditor from "react-ace";
+import * as Modal from "react-modal";
+
+import "brace/mode/json";
+import "brace/mode/ruby";
+import "brace/mode/text";
+
+import { IFunction } from "../../shared/types";
+
+// TODO: fetch this from the API/ConfigMap
+const Runtimes = {
+  "Nodejs (6)": "nodejs6",
+  "Nodejs (8)": "nodejs8",
+  "Python (2.7)": "python2.7",
+  "Python (3.4)": "python3.4",
+  "Ruby (2.4)": "ruby2.4",
+};
+
+interface IFunctionDeployButtonProps {
+  deployFunction: (n: string, ns: string, spec: IFunction["spec"]) => Promise<any>;
+  navigateToFunction: (n: string, ns: string) => Promise<any>;
+}
+
+interface IFunctionDeployButtonState {
+  functionSpec: IFunction["spec"];
+  modalIsOpen: boolean;
+  name: string;
+  namespace: string;
+  error?: string;
+}
+
+class FunctionDeployButton extends React.Component<
+  IFunctionDeployButtonProps,
+  IFunctionDeployButtonState
+> {
+  public state: IFunctionDeployButtonState = {
+    functionSpec: {
+      deps: "",
+      function: "",
+      handler: "hello.handler",
+      runtime: "nodejs6",
+      type: "HTTP",
+    },
+    modalIsOpen: false,
+    name: "",
+    namespace: "default",
+  };
+
+  public componentDidMount() {
+    const generatedName = Moniker.choose();
+    this.setState({
+      functionSpec: {
+        ...this.state.functionSpec,
+        function: this.generateFunction(
+          this.state.functionSpec.runtime,
+          this.state.functionSpec.handler,
+        ),
+        handler: `${generatedName}.handler`,
+      },
+      name: generatedName,
+    });
+  }
+
+  public render() {
+    const { functionSpec: f, name, namespace } = this.state;
+    return (
+      <div className="FunctionDeployButton">
+        <button className="button button-accent" onClick={this.openModal}>
+          Deploy New Function
+        </button>
+        <Modal
+          isOpen={this.state.modalIsOpen}
+          onRequestClose={this.closeModal}
+          contentLabel="Modal"
+        >
+          {this.state.error && (
+            <div className="container padding-v-big margin-v-big bg-action">{this.state.error}</div>
+          )}
+          <form onSubmit={this.handleDeploy}>
+            <div>
+              <label htmlFor="name">Name</label>
+              <input id="name" onChange={this.handleNameChange} value={name} required={true} />
+            </div>
+            <div>
+              <label htmlFor="namespace">Namespace</label>
```

Comment:
I am curious: Do you support changing the namespace in the view? What happens if you create the namespace different than the default?

Reply:
We do, if the namespace doesn't exist it'll return an error, so it expects an existing namespace. We don't want to give the ServiceAccount the ability to create namespaces, when we have user authz we can try to create the namespace.

Reply:
cool